File: server/src/main/java/com/example/service/WorldService.java
package com.example.service;

import com.example.domain.Campaign;
import com.example.domain.Location;
import com.example.domain.PointOfInterest;
import com.example.domain.PointOfInterestCategory;
import com.example.domain.PartyLocation;
import com.example.domain.Region;
import com.example.domain.Settlement;
import com.example.domain.SettlementType;
import com.example.domain.TravelRoute;
import com.example.db.CampaignRepository;
import com.example.db.LocationRepository;
import com.example.db.PointOfInterestRepository;
import com.example.db.PartyLocationRepository;
import com.example.db.RegionRepository;
import com.example.db.SettlementRepository;
import com.example.db.TravelRouteRepository;
import com.example.dto.TravelPlan;
import com.example.service.CampaignEventService;

import org.springframework.stereotype.Service;

import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.PriorityQueue;

/**
 * Business and world logic for the world model owned by a campaign: regions,
 * locations, settlements, points of interest, travel routes, and the party's current
 * location.
 *
 * <p>This service is the single place where the world is created, explored, and
 * traversed. Every mutation resolves its owning campaign, applies the change to a
 * managed entity, and relies on the repositories to persist it, so the world reloads
 * across sessions. The {@link Location} is the canonical place node: settlements and
 * points of interest each reference one, routes connect two, and the party location
 * points at one.</p>
 */
@Service
public class WorldService {

    private final CampaignRepository campaigns;
    private final RegionRepository regions;
    private final LocationRepository locations;
    private final SettlementRepository settlements;
    private final PointOfInterestRepository pointsOfInterest;
    private final TravelRouteRepository travelRoutes;
    private final PartyLocationRepository partyLocations;
    private final CampaignEventService events;
    private final DtoValidator validator;

    public WorldService(
            CampaignRepository campaigns,
            RegionRepository regions,
            LocationRepository locations,
            SettlementRepository settlements,
            PointOfInterestRepository pointsOfInterest,
            TravelRouteRepository travelRoutes,
            PartyLocationRepository partyLocations,
            CampaignEventService events,
            DtoValidator validator) {
        this.campaigns = campaigns;
        this.regions = regions;
        this.locations = locations;
        this.settlements = settlements;
        this.pointsOfInterest = pointsOfInterest;
        this.travelRoutes = travelRoutes;
        this.partyLocations = partyLocations;
        this.events = events;
        this.validator = validator;
    }

    // ------------------------------------------------------------------
    // Campaign lookup
    // ------------------------------------------------------------------

    private Campaign requireCampaign(Long campaignId) {
        return campaigns.findById(campaignId)
                .orElseThrow(() -> new IllegalArgumentException("No campaign with id " + campaignId));
    }

    // ------------------------------------------------------------------
    // Regions
    // ------------------------------------------------------------------

    public Region addRegion(Long campaignId, String name, String description) {
        Campaign campaign = requireCampaign(campaignId);
        return regions.save(new Region(campaign, name, description));
    }

    public List<Region> listRegions(Long campaignId) {
        Campaign campaign = requireCampaign(campaignId);
        return regions.findByCampaignOrderByName(campaign);
    }

    // ------------------------------------------------------------------
    // Locations
    // ------------------------------------------------------------------

    public Location addLocation(Long campaignId, String name, String description,
                                Long regionId, Double latitude, Double longitude) {
        Campaign campaign = requireCampaign(campaignId);
        Location location = new Location(campaign, name, description, false);
        if (regionId != null) {
            location.setRegion(regions.findById(regionId)
                    .orElseThrow(() -> new IllegalArgumentException("No region with id " + regionId)));
        }
        location.setLatitude(latitude);
        location.setLongitude(longitude);
        return locations.save(location);
    }

    public List<Location> listLocations(Long campaignId) {
        requireCampaign(campaignId);
        return locations.findByCampaignOrderByName(requireCampaign(campaignId));
    }

    /**
     * Marks a location discovered. Undiscovered locations are hidden from views that
     * filter by discovery, so this is what turns a place from "unknown" into "known".
     */
    public Location discoverLocation(Long campaignId, Long locationId) {
        Location location = requireLocation(campaignId, locationId);
        boolean wasUndiscovered = !location.isDiscovered();
        location.discover();
        locations.save(location);
        if (wasUndiscovered) {
            events.recordDiscovery(campaignId, location.getName(), location.getId(), "location");
        }
        return location;
    }

    public List<Location> listDiscoveredLocations(Long campaignId) {
        requireCampaign(campaignId);
        return locations.findByCampaignIdAndDiscovered(campaignId, true);
    }

    // ------------------------------------------------------------------
    // Settlements
    // ------------------------------------------------------------------

    public Settlement addSettlement(Long campaignId, String name, String description,
                                    SettlementType type, int population, Long regionId) {
        validator.requireNonNegativeQuantity("Settlement population", population);
        Location location = addLocation(campaignId, name, description, regionId, null, null);
        Campaign campaign = requireCampaign(campaignId);
        return settlements.save(new Settlement(campaign, location, type, population));
    }

    public List<Settlement> listSettlements(Long campaignId) {
        requireCampaign(campaignId);
        return settlements.findByCampaign(requireCampaign(campaignId));
    }

    // ------------------------------------------------------------------
    // Points of interest
    // ------------------------------------------------------------------

    public PointOfInterest addPointOfInterest(Long campaignId, String name, String description,
                                              PointOfInterestCategory category,
                                              Long settlementId, Long regionId) {
        Location location = addLocation(campaignId, name, description, regionId, null, null);
        Campaign campaign = requireCampaign(campaignId);
        Settlement settlement = null;
        if (settlementId != null) {
            settlement = settlements.findById(settlementId)
                    .orElseThrow(() -> new IllegalArgumentException("No settlement with id " + settlementId));
        }
        return pointsOfInterest.save(new PointOfInterest(campaign, location, category, settlement));
    }

    public List<PointOfInterest> listPointsOfInterest(Long campaignId) {
        requireCampaign(campaignId);
        return pointsOfInterest.findByCampaign(requireCampaign(campaignId));
    }

    // ------------------------------------------------------------------
    // Travel routes
    // ------------------------------------------------------------------

    public TravelRoute addTravelRoute(Long campaignId, Long fromId, Long toId,
                                      Double distanceKm, Integer travelMinutes) {
        validator.requirePositiveIdentifier("location", fromId);
        validator.requirePositiveIdentifier("location", toId);
        validator.requireNonNegativeQuantity("Travel distance (km)", distanceKm);
        validator.requireNonNegativeQuantity("Travel time (minutes)", travelMinutes);
        Campaign campaign = requireCampaign(campaignId);
        Location from = requireLocation(campaignId, fromId);
        Location to = requireLocation(campaignId, toId);
        return travelRoutes.save(new TravelRoute(campaign, from, to, distanceKm, travelMinutes));
    }

    public List<TravelRoute> listTravelRoutes(Long campaignId) {
        requireCampaign(campaignId);
        return travelRoutes.findByCampaign(requireCampaign(campaignId));
    }

    public void removeTravelRoute(Long routeId) {
        travelRoutes.findById(routeId)
                .orElseThrow(() -> new IllegalArgumentException("No travel route with id " + routeId));
        travelRoutes.deleteById(routeId);
    }

    // ------------------------------------------------------------------
    // Current party location
    // ------------------------------------------------------------------

    public PartyLocation setPartyLocation(Long campaignId, Long locationId) {
        validator.requirePositiveIdentifier("location", locationId);
        Location location = requireLocation(campaignId, locationId);
        PartyLocation partyLocation = partyLocations.findByCampaignId(campaignId)
                .orElseGet(() -> new PartyLocation(campaignId, location));
        boolean moved = partyLocation.getLocation() == null
                || !location.getId().equals(partyLocation.getLocation().getId());
        partyLocation.setLocation(location);
        partyLocations.save(partyLocation);
        if (moved) {
            events.recordLocationEntry(campaignId, location.getId(), location.getName());
        }
        return partyLocation;
    }

    public Optional<PartyLocation> findPartyLocation(Long campaignId) {
        return partyLocations.findByCampaignId(campaignId);
    }

    // ------------------------------------------------------------------
    // Travel planning
    // ------------------------------------------------------------------

    /**
     * Plans the cheapest journey between two locations in a campaign by travelling
     * through the directed {@link TravelRoute} graph. Cheapest is measured in travel
     * time; the associated distance is summed along the chosen path.
     *
     * <p>The route graph is rebuilt from the campaign's routes on every call so that
     * freshly created or removed routes are reflected immediately. When no sequence of
     * routes connects the origin and the destination, the returned plan is marked
     * {@link TravelPlan#reachable() unreachable}.</p>
     *
     * @param campaignId the owning campaign
     * @param fromId     the origin location id
     * @param toId       the destination location id
     * @return the travel plan (never {@code null})
     */
    public TravelPlan computeTravel(Long campaignId, Long fromId, Long toId) {
        requireCampaign(campaignId);
        requireLocation(campaignId, fromId);
        requireLocation(campaignId, toId);

        if (fromId.equals(toId)) {
            return new TravelPlan(fromId, toId, true, 0, 0.0);
        }

        Map<Long, Map<Long, RouteEdge>> adjacency = buildAdjacency(campaignId);
        return dijkstra(adjacency, fromId, toId);
    }

    private Map<Long, Map<Long, RouteEdge>> buildAdjacency(Long campaignId) {
        Map<Long, Map<Long, RouteEdge>> adjacency = new HashMap<>();
        for (TravelRoute route : travelRoutes.findByCampaign(requireCampaign(campaignId))) {
            adjacency.computeIfAbsent(route.getFrom().getId(), k -> new HashMap<>())
                    .put(route.getTo().getId(), new RouteEdge(route));
        }
        return adjacency;
    }

    private TravelPlan dijkstra(
            Map<Long, Map<Long, RouteEdge>> adjacency, Long fromId, Long toId) {
        Set<Long> visited = new HashSet<>();
        PriorityQueue<RouteEdge> frontier =
                new PriorityQueue<>(Comparator.comparingInt(e -> e.minutes));
        frontier.add(new RouteEdge(fromId, toId, 0, 0.0));

        while (!frontier.isEmpty()) {
            RouteEdge edge = frontier.poll();
            if (visited.contains(edge.to)) {
                continue;
            }
            visited.add(edge.to);
            if (edge.to.equals(toId)) {
                return new TravelPlan(fromId, toId, true, edge.minutes, edge.distance);
            }
            Map<Long, RouteEdge> neighbours = adjacency.get(edge.to);
            if (neighbours != null) {
                for (RouteEdge next : neighbours.values()) {
                    if (!visited.contains(next.to)) {
                        frontier.add(next.withAccumulated(edge.minutes, edge.distance));
                    }
                }
            }
        }

        return TravelPlan.unreachable(fromId, toId);
    }

    /**
     * A single directed step in the travel graph: one {@link TravelRoute} or a
     * synthetic starting step with zero accumulated cost.
     */
    private static final class RouteEdge {
        private final Long from;
        private final Long to;
        private final Integer minutes;
        private final Double distance;

        RouteEdge(TravelRoute route) {
            this(route.getFrom().getId(), route.getTo().getId(),
                    route.getTravelMinutes() == null ? 0 : route.getTravelMinutes(),
                    route.getDistanceKm() == null ? 0.0 : route.getDistanceKm());
        }

        RouteEdge(Long from, Long to, Integer minutes, Double distance) {
            this.from = from;
            this.to = to;
            this.minutes = minutes;
            this.distance = distance;
        }

        RouteEdge withAccumulated(Integer accumulatedMinutes, Double accumulatedDistance) {
            return new RouteEdge(from, to,
                    accumulatedMinutes + minutes, accumulatedDistance + distance);
        }
    }

    // ------------------------------------------------------------------
    // Lookups
    // ------------------------------------------------------------------

    private Location requireLocation(Long campaignId, Long locationId) {
        return locations.findById(locationId)
                .filter(l -> l.getCampaign() != null && l.getCampaign().getId() != null
                        && l.getCampaign().getId().equals(campaignId))
                .orElseThrow(() -> new IllegalArgumentException("No location with id " + locationId));
    }
}

File: client/src/app/services/world.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import {
  CreateLocationRequest,
  CreatePointOfInterestRequest,
  CreateRegionRequest,
  CreateSettlementRequest,
  Location,
  PointOfInterest,
  PointOfInterestCategory,
  Region,
  Settlement,
  SettlementType,
} from '../models/world';

/**
 * The front-end client for the back-end world system.
 *
 * <p>This service wraps {@code /api/campaigns/{campaignId}/...} world endpoints: regions,
 * locations, settlements, points of interest, travel routes, and the party's current location.
 * World mutations use request parameters, so this service builds typed {@link HttpParams} for them.</p>
 *
 * <p>The {@link CampaignStore} owns the authoritative world data for the active campaign; this
 * service only performs the HTTP round trips.</p>
 */
@Injectable({ providedIn: 'root' })
export class WorldService {
  private readonly http = inject(HttpClient);

  // ------------------------------------------------------------------
  // Regions
  // ------------------------------------------------------------------

  /**
   * @param campaignId the owning campaign
   * @return every region in the campaign
   */
  listRegions(campaignId: number): Observable<Region[]> {
    return this.http.get<Region[]>(`/api/campaigns/${campaignId}/regions`);
  }

  /**
   * Creates a new region.
   *
   * @param campaignId the owning campaign
   * @param request the region creation request
   * @return the created region
   */
  createRegion(campaignId: number, request: CreateRegionRequest): Observable<Region> {
    return this.http.post<Region>(
      `/api/campaigns/${campaignId}/regions`,
      {},
      { params: this.params(request) },
    );
  }

  // ------------------------------------------------------------------
  // Locations
  // ------------------------------------------------------------------

  /**
   * @param campaignId the owning campaign
   * @return every location in the campaign
   */
  listLocations(campaignId: number): Observable<Location[]> {
    return this.http.get<Location[]>(`/api/campaigns/${campaignId}/locations`);
  }

  /**
   * @param campaignId the owning campaign
   * @return only the discovered locations in the campaign
   */
  listDiscoveredLocations(campaignId: number): Observable<Location[]> {
    return this.http.get<Location[]>(
      `/api/campaigns/${campaignId}/locations/discovered`,
    );
  }

  /**
   * Creates a new location.
   *
   * @param campaignId the owning campaign
   * @param request the location creation request
   * @return the created location
   */
  createLocation(
    campaignId: number,
    request: CreateLocationRequest,
  ): Observable<Location> {
    return this.http.post<Location>(
      `/api/campaigns/${campaignId}/locations`,
      {},
      { params: this.params(request) },
    );
  }

  /**
   * Marks a location as discovered.
   *
   * @param campaignId the owning campaign
   * @param locationId the location to discover
   * @return the updated location
   */
  discover(campaignId: number, locationId: number): Observable<Location> {
    return this.http.post<Location>(
      `/api/campaigns/${campaignId}/locations/${locationId}/discover`,
      null,
    );
  }

  // ------------------------------------------------------------------
  // Settlements
  // ------------------------------------------------------------------

  /**
   * @param campaignId the owning campaign
   * @return every settlement in the campaign
   */
  listSettlements(campaignId: number): Observable<Settlement[]> {
    return this.http.get<Settlement[]>(
      `/api/campaigns/${campaignId}/settlements`,
    );
  }

  /**
   * Creates a new settlement.
   *
   * @param campaignId the owning campaign
   * @param request the settlement creation request
   * @return the created settlement
   */
  createSettlement(
    campaignId: number,
    request: CreateSettlementRequest,
  ): Observable<Settlement> {
    let params = new HttpParams()
      .append('name', request.name)
      .append('type', request.type ?? SettlementType.VILLAGE)
      .append('population', String(request.population ?? 0));
    if (request.regionId !== undefined && request.regionId !== null) {
      params = params.append('regionId', String(request.regionId));
    }
    return this.http.post<Settlement>(
      `/api/campaigns/${campaignId}/settlements`,
      {},
      { params },
    );
  }

  // ------------------------------------------------------------------
  // Points of interest
  // ------------------------------------------------------------------

  /**
   * Creates a new point of interest. The back-end creates the backing location for the
   * point of interest, so the caller only supplies the name, category and description.
   *
   * @param campaignId the owning campaign
   * @param request the point-of-interest creation request
   * @return the created point of interest
   */
  createPointOfInterest(
    campaignId: number,
    request: CreatePointOfInterestRequest,
  ): Observable<PointOfInterest> {
    let params = new HttpParams()
      .append('name', request.name)
      .append('category', request.category ?? PointOfInterestCategory.LANDMARK);
    if (request.description) {
      params = params.append('description', request.description);
    }
    return this.http.post<PointOfInterest>(
      `/api/campaigns/${campaignId}/points-of-interest`,
      {},
      { params },
    );
  }

  // ------------------------------------------------------------------
  // Points of interest
  // ------------------------------------------------------------------

  /**
   * @param campaignId the owning campaign
   * @return every point of interest in the campaign
   */
  listPointsOfInterest(campaignId: number): Observable<PointOfInterest[]> {
    return this.http.get<PointOfInterest[]>(
      `/api/campaigns/${campaignId}/points-of-interest`,
    );
  }

  // ------------------------------------------------------------------
  // Party location
  // ------------------------------------------------------------------

  /**
   * @param campaignId the owning campaign
   * @return the party's current location, if set
   */
  getPartyLocation(campaignId: number): Observable<Partial<{ locationId: number }>> {
    return this.http.get<Partial<{ locationId: number }>>(
      `/api/campaigns/${campaignId}/party-location`,
    );
  }

  /**
   * Sets the party's current location.
   *
   * @param campaignId the owning campaign
   * @param locationId the location the party has moved to
   * @return the updated party location
   */
  setPartyLocation(campaignId: number, locationId: number): Observable<unknown> {
    return this.http.post<unknown>(
      `/api/campaigns/${campaignId}/party-location`,
      {},
      { params: new HttpParams().append('locationId', String(locationId)) },
    );
  }

  /**
   * Builds request parameters from an object, skipping undefined and null values.
   */
  private params(fields: object): HttpParams {
    let params = new HttpParams();
    for (const [key, value] of Object.entries(fields)) {
      if (value !== undefined && value !== null) {
        params = params.append(key, String(value));
      }
    }
    return params;
  }
}
